control_apps/src/control_mngr.rs

```diff
@@ -0,0 +1,292 @@
+            self.valve_dump.set_pwm(0.0);
+            self.valve_dump.lock();
+            // enable venting
+            self.valve_vent.unlock();
+            self.mode = ControlMode::Vent;
+            info!("CtrlMode transition: {:} --> {:}", previous_mode, self.mode);
+        } else {
+            warn!(
+                "CtrlMode transition ({:} --> {:}) not allowed! Ignoring...",
+                previous_mode,
+                ControlMode::Idle
+            );
+        }
+    }
+
+    fn dump(&mut self) {
+        // enter mode used for raising altitude
+        debug!("Attempting transition to {:}", ControlMode::Dump);
+        let previous_mode = self.mode;
+        if previous_mode == ControlMode::Idle || previous_mode == ControlMode::Vent {
+            // stop vemtomg -- can only do one at a time!
+            self.valve_vent.set_pwm(0.0);
+            self.valve_vent.lock();
+            // enable dumping
+            self.valve_dump.unlock();
+            self.mode = ControlMode::Dump;
+            info!("CtrlMode transition: {:} --> {:}", previous_mode, self.mode);
+        } else {
+            warn!(
+                "CtrlMode transition ({:} --> {:}) not allowed! Ignoring...",
+                previous_mode,
+                ControlMode::Idle
+            );
+        }
+    }
+
+    pub fn abort(&mut self) {
+        // dump all ballast and lock balloon valve closed
+        debug!("Attempting transition to {:}", ControlMode::Abort);
+        let previous_mode = self.mode;
+        // unlock the valves if they are locked to force the next step!
+        self.valve_vent.unlock();
+        self.valve_dump.unlock();
+        // execute mode transition
+        self.mode = ControlMode::Abort;
+        warn!("CtrlMode transition: {:} --> {:}", previous_mode, self.mode);
+        // lock the vent valve closed
+        self.valve_vent.set_pwm(0.0);
+        self.valve_vent.lock();
+        // lock the dump valve open
+        self.valve_dump.set_pwm(1.0);
+        self.valve_dump.lock();
+    }
+
+    pub fn get_mode(&self) -> ControlMode {
+        return self.mode;
+    }
+
+    pub fn set_target(&mut self, target_altitude: f32) {
+        // set new target altitude
+        // Set a new target altitude to converge toward (in meters)
+        if target_altitude > CTRL_ALTITUDE_FLOOR {
+            // target must be above the minimum allowed altitude
+            self.target_altitude = target_altitude;
+            info!("New target altitude: {:}m", self.target_altitude);
+        } else {
+            warn!(
+                "Not allowed to set target altitude below {:}! Ignoring...",
+                CTRL_ALTITUDE_FLOOR
+            );
+        }
+    }
+
+    pub fn start_control(&mut self) {
+        // enable altitude control+transition to idle
+        let previous_mode = self.get_mode();
+        if previous_mode == ControlMode::Safe || previous_mode == ControlMode::Idle {
```

Comment:
I don't think you would need to care about the previous mode if you used a more structured way to perform this state machine.

If a state behaves differently depending on what the previous state was, then you have more than one state instead!